### PR TITLE
Beta History scrolling behavior and polling efficiency tuning

### PR DESCRIPTION
**Pull request by Nerdinacan**
## What did you do? 
- Reworked content providers in the beta history to smooth out the polling and cache delivery cycles. This is a complex process and the initial release has a lot of flaws that were noted by testers.
- ~~I've simplified the server-side query to work on limit/offset rather than a bi-direction query. In order to make this all work we wait for the first poll to the server to establish a totalMatches value as well as some other statistics available only from the server. Those numbers make it easier to later render the filtered cache results. So we trade one mandatory ajax query for the counts (which was going to be made anyway) for a content query endpoint that only has to do half as much work on each polling request.~~ Gah I knew that was too good to be true. We can't omit the bi-directional query because of the subsequent updates during polling. During polling, we further filter each poll request with an update_time parameter making the limit/offset refer to the wrong rows. I think we're stuck with the bi-directional query on the server. 
- The big challenge here is making the client look like it has all the same information as the server, even though it obviously can't. So the cache-side query will never be as simple as the server side query simply because there is no way to exactly know the key totals necessary for rendering the scroller (for an arbitrary selection), therefore some of those values must be estimated. But I'm attempting to explain how that happens in the documentation.
- I'm adding documentation to explain how the providers work including flowcharts for the RxJS observables. I've heard rumors nobody understands how the RxJS works which is unfortunate because I don't think Vue's observability system can give us everything we need on its own, so I'm attempting to simplify and document what's actually happening with those streams.
- Added a new scroller component that allows for both fine-tuning of nearby results (via the mouse wheel) and bigger switches between regions (via scrollbar). This is necessary because with a large history, say 1 million rows, divided by ~1200 vertical pixels in the average screen, you end up with about 1000 rows per pixel meaning that even the tiniest mousewheel of scroll will never get to the row you care about. Now the mousewheel moves a few rows at a time but the scrollbar moves proportionately (by comparing the percentage of the scroll bar to the entire result set length).

## Why did you make this change?
https://github.com/galaxyproject/galaxy/issues/10947

Also I just didn't like how clunky it felt regardless of what other people said. :)

## How to test the changes? 
(select the most appropriate option; if the latter, provide steps for testing below)
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [x] This is a refactoring of components with existing test coverage.
- [x] Instructions for manual testing are as follows:
  1. Coming, still working...

## For UI Components
- [ ] I've included a screenshot of the changes

### CODE REVIEW

**Comment by mvdbeek** [comment]
> * Added a new scroller component that allows for both fine-tuning of nearby results (via the mouse wheel) and bigger switches between regions (via scrollbar). This is necessary because with a large history, say 1 million rows, divided by ~1200 vertical pixels in the average screen, you end up with about 1000 rows per pixel meaning that even the tiniest mousewheel of scroll will never get to the row you care about. Now the mousewheel moves a few rows at a time but the scrollbar moves proportionately (by comparing the percentage of the scroll bar to the entire result set length).

💯 that is awesome and I think this is nice and intuitive!

> * Reworked content providers in the beta history to smooth out the polling and cache delivery cycles. This is a complex process and the initial release has a lot of flaws that were noted by testers.

Navigating out of a collection and switching histories is now snappy enough, that is great!

When you go into a collection the loading indicator never disappears.
Screenshot 2021-04-13 at 09 32 02

Looks like you're close!

**Comment by Nerdinacan** [comment]
> > * Added a new scroller component that allows for both fine-tuning of nearby results (via the mouse wheel) and bigger switches between regions (via scrollbar). This is necessary because with a large history, say 1 million rows, divided by ~1200 vertical pixels in the average screen, you end up with about 1000 rows per pixel meaning that even the tiniest mousewheel of scroll will never get to the row you care about. Now the mousewheel moves a few rows at a time but the scrollbar moves proportionately (by comparing the percentage of the scroll bar to the entire result set length).
> 
> 💯 that is awesome and I think this is nice and intuitive!
> 
> > * Reworked content providers in the beta history to smooth out the polling and cache delivery cycles. This is a complex process and the initial release has a lot of flaws that were noted by testers.
> 
> Navigating out of a collection and switching histories is now snappy enough, that is great!
> 
> When you go into a collection the loading indicator never disappears.
> Screenshot 2021-04-13 at 09 32 02
> 
> Looks like you're close!

Got it. I used to have a more sophisticated loader detection, but I scrapped it in favor of this simple on/off mechanism. I was just forgetting to manually update the output subject there.

**Comment by Nerdinacan** [comment]
> > * Added a new scroller component that allows for both fine-tuning of nearby results (via the mouse wheel) and bigger switches between regions (via scrollbar). This is necessary because with a large history, say 1 million rows, divided by ~1200 vertical pixels in the average screen, you end up with about 1000 rows per pixel meaning that even the tiniest mousewheel of scroll will never get to the row you care about. Now the mousewheel moves a few rows at a time but the scrollbar moves proportionately (by comparing the percentage of the scroll bar to the entire result set length).
> 
> 💯 that is awesome and I think this is nice and intuitive!
> 
> > * Reworked content providers in the beta history to smooth out the polling and cache delivery cycles. This is a complex process and the initial release has a lot of flaws that were noted by testers.
> 
> Navigating out of a collection and switching histories is now snappy enough, that is great!
> 
> When you go into a collection the loading indicator never disappears.
> Screenshot 2021-04-13 at 09 32 02
> 
> Looks like you're close!

@mvdbeek Thanks for looking man. As usual, it was way more work than expected.

**Comment by mvdbeek** [comment]
One more thing, is it possible to stop the scroller once you've definitely reached the first dataset ?
In this history
Screenshot 2021-04-13 at 17 12 06
 a light scroll brings me to
Screenshot 2021-04-13 at 17 11 58

Ideally the scroller would stop once you've reach the first hid, so you can't scroll further down. It's not obvious that there are datasets if you scroll up. Happy to approve without that though, this is a big step forward already.

**Comment by Nerdinacan** [comment]
> Ideally the scroller would stop once you've reach the first hid, so you can't scroll further down. It's not obvious that there are datasets if you scroll up. Happy to approve without that though, this is a big step forward already.

@mvdbeek Believe it or not, that request is harder than you might expect. I'll take a peek at it though. What you're asking for is certainly what I tried to deliver first, and is the expected behavior of a scroller, it's just this custom scroller is missing a lot of key pieces of information that are normally available to an HTML element that already knows... for instance, how much stuff is going to end up in the list.

**Review by mvdbeek — APPROVED** [review]

**Comment by Nerdinacan** [comment]
Ahhh I just noticed a problem with collections. Don't merge yet!

**Comment by Nerdinacan** [comment]
That multi-history selenium test is failing even on unrelated PRs, I do not believe it is related to these changes.

File: client/src/components/History/CurrentCollection/Panel.vue [hunk 1]
```diff
@@ -25,13 +18,21 @@
...
+                            <input
+                                type="range"
+                                min="0.0"
+                                max="1.0"
+                                step="0.1"
+                                :value="scrollPos.cursor"
+                                @input="setScrollPos({ cursor: 1.0 * $event.srcElement.value })"
+                            />
+                            <pre>{{ payload | reportPayload("element_index") }}</pre>
+                        </div> -->
```
Comment (mvdbeek) [inline client/src/components/History/CurrentCollection/Panel.vue]:
```suggestion
```
That's just leftover from debugging, right ?

File: client/src/components/History/History.vue [hunk 2]
```diff
@@ -43,32 +42,38 @@
...
+                            <input
+                                type="range"
+                                min="0.0"
+                                max="1.0"
+                                step="0.1"
+                                :value="scrollPos.cursor"
+                                @input="setScrollPos({ cursor: 1.0 * $event.srcElement.value })"
+                            />
+                            <pre>{{ payload | reportPayload }}</pre>
+                        </div> -->
```
Comment (mvdbeek) [inline client/src/components/History/History.vue]:
```suggestion
```

File: client/src/components/History/providers/CollectionContentProvider/CollectionContentProvider.js [hunk 3]
```diff
@@ -32,18 +17,19 @@ export default {
     watch: {
         dsc(newDsc, oldDsc) {
             if (!(newDsc.id == oldDsc.id)) {
+                console.log("resetting scroll pos");
```
Comment (mvdbeek) [inline client/src/components/History/providers/CollectionContentProvider/CollectionContentProvider.js]:
```suggestion
```

**Review by mvdbeek — COMMENTED** [review]
Still WIP ?

**Comment by Nerdinacan** [comment]
@mvdbeek It's not still WIP, sorry I forgot that tag was on it

**Pull request merged**